File: backend/apps/sales/pricing.py
"""
What a sale costs.

Pure arithmetic, no database. Everything here is integer cents, and the rules
are the ones already proved in milestone 2: half-up rounding, tax rates in basis
points, and inclusive tax derived by subtraction so net plus tax is always
exactly what the customer pays.

Three rules govern the order of operations, and each exists because getting it
the other way round produces a receipt that is wrong in a way a customer can
see:

**Discount before tax.** A discount reduces the taxable amount. Taxing first and
discounting after would charge tax on money nobody paid.

**A cart discount is apportioned across lines before tax is computed.** Tax is
per line, because each line carries its own rate and its own inclusive flag, so
there is no single rate at which a whole-cart discount could be taxed. Leaving
it unapportioned would make the tax total unreconcilable with the lines.

**Round per line, then sum.** Never compute a total and round that. Rounding the
sum lets a total disagree with the lines printed above it, which is precisely
the thing a customer checks.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from decimal import Decimal

from apps.core.money import apply_percentage, split_exclusive, split_inclusive


class PricingError(ValueError):
    """A cart that cannot be priced, as opposed to one that prices to zero."""


@dataclass(frozen=True)
class LineInput:
    """One line as the till submits it."""

    item_id: str
    name: str
    sku: str
    unit: str
    unit_price_cents: int
    quantity: Decimal
    tax_rate_bps: int = 0
    tax_is_inclusive: bool = True
    #: Line discount, given either way. Both may be present; both are applied.
    discount_bps: int = 0
    discount_cents: int = 0

    def gross_before_discount(self) -> int:
        """Quantity times price, rounded once.

        Rounded here because quantity is fractional for goods sold by weight -
        2.5 kg at KES 180 is 450, but 0.333 kg is 59.94, and the cent has to be
        resolved somewhere. Doing it once, at the line, keeps every later figure
        exact.
        """
        exact = Decimal(self.unit_price_cents) * self.quantity
        return int(exact.to_integral_value(rounding="ROUND_HALF_UP"))


@dataclass
class LineTotals:
    """One priced line."""

    line: LineInput
    gross_before_discount_cents: int
    line_discount_cents: int
    cart_discount_share_cents: int
    net_cents: int
    tax_cents: int
    gross_cents: int

    @property
    def total_discount_cents(self) -> int:
        return self.line_discount_cents + self.cart_discount_share_cents


@dataclass
class CartTotals:
    """A whole priced cart."""

    lines: list[LineTotals] = field(default_factory=list)
    subtotal_cents: int = 0
    discount_cents: int = 0
    tax_cents: int = 0
    total_cents: int = 0

    @property
    def line_count(self) -> int:
        return len(self.lines)


def resolve_discount(base_cents: int, *, discount_bps: int, discount_cents: int) -> int:
    """Work out a discount in cents, however it was expressed.

    Percentage and fixed amount may both be present, and both apply - a "10% off
    and another 50 shillings" promotion is an ordinary thing to want. The result
    is capped at the base, because a discount larger than the thing being
    discounted would produce a negative line and, downstream, negative tax.
    """
    if discount_bps < 0 or discount_cents < 0:
        raise PricingError("A discount cannot be negative.")
    if discount_bps > 10_000:
        raise PricingError("A discount cannot exceed 100%.")

    total = apply_percentage(base_cents, discount_bps) + discount_cents
    return min(total, max(base_cents, 0))


def apportion(amount_cents: int, weights: list[int]) -> list[int]:
    """Split an amount across weights so the parts sum exactly to the whole.

    Largest-remainder: each weight takes its exact share floored, and the cents
    left over go to whichever lines were rounded down hardest, ties broken by
    the larger weight so the result is deterministic rather than dependent on
    dictionary order.

    The exactness is the point. A cart discount that apportions to one cent less
    than it should leaves a total that does not match the sum of its lines, and
    that discrepancy would then be silently absorbed into tax.
    """
    if amount_cents == 0 or not weights:
        return [0] * len(weights)
    if any(weight < 0 for weight in weights):
        raise PricingError("Weights cannot be negative.")

    total_weight = sum(weights)
    if total_weight == 0:
        # Every line is free, so there is nothing to take a discount off. The
        # alternative - splitting equally - would create negative lines.
        return [0] * len(weights)

    shares = [amount_cents * weight // total_weight for weight in weights]
    remainder = amount_cents - sum(shares)

    if remainder:
        # Rank by the fractional part that was discarded, largest first.
        fractions = [
            (amount_cents * weight) % total_weight
            for weight in weights
        ]
        order = sorted(
            range(len(weights)),
            key=lambda index: (-fractions[index], -weights[index], index),
        )
        for position in range(remainder):
            shares[order[position % len(order)]] += 1

    return shares


def price_cart(
    lines: list[LineInput],
    *,
    cart_discount_bps: int = 0,
    cart_discount_cents: int = 0,
) -> CartTotals:
    """Price a whole cart, lines and all.

    The sequence is: line gross, line discount, apportion the cart discount over
    what remains, then split tax on each line using that line's own rate and
    inclusive flag. Mixed inclusive and exclusive lines therefore total
    correctly on one sale, which is the case a per-business tax setting could
    not express.
    """
    if not lines:
        raise PricingError("A sale needs at least one line.")

    grosses = [line.gross_before_discount() for line in lines]
    line_discounts = [
        resolve_discount(
            gross, discount_bps=line.discount_bps, discount_cents=line.discount_cents
        )
        for line, gross in zip(lines, grosses, strict=True)
    ]
    after_line_discount = [
        gross - discount for gross, discount in zip(grosses, line_discounts, strict=True)
    ]

    cart_discount = resolve_discount(
        sum(after_line_discount),
        discount_bps=cart_discount_bps,
        discount_cents=cart_discount_cents,
    )
    shares = apportion(cart_discount, after_line_discount)

    totals = CartTotals()
    for line, gross, line_discount, share, base in zip(
        lines, grosses, line_discounts, shares, after_line_discount, strict=True
    ):
        charged = base - share

        if line.tax_rate_bps <= 0:
            net, tax = charged, 0
            line_gross = charged
        elif line.tax_is_inclusive:
            # The amount charged already contains the tax.
            net, tax = split_inclusive(charged, line.tax_rate_bps)
            line_gross = charged
        else:
            # Tax is added on top of the amount charged.
            line_gross, tax = split_exclusive(charged, line.tax_rate_bps)
            net = charged

        totals.lines.append(
            LineTotals(
                line=line,
                gross_before_discount_cents=gross,
                line_discount_cents=line_discount,
                cart_discount_share_cents=share,
                net_cents=net,
                tax_cents=tax,
                gross_cents=line_gross,
            )
        )

        totals.subtotal_cents += net
        totals.discount_cents += line_discount + share
        totals.tax_cents += tax
        totals.total_cents += line_gross

    return totals


def change_for(tendered_cents: int, due_cents: int) -> int:
    """Change owed on a cash tender."""
    if tendered_cents < due_cents:
        raise PricingError("Tendered less than the amount due.")
    return tendered_cents - due_cents

File: mobile/lib/data/cart/pricing.dart
/// What a cart costs, computed on the till.
///
/// **This is a deliberate second implementation of `apps/sales/pricing.py`, and
/// it has to agree with it exactly.** Normally two implementations of the same
/// arithmetic is a mistake. Here the till has to price a cart with no
/// connection - a cashier cannot be told "no total until the network comes
/// back" while a customer waits - so the choice is between this and no offline
/// selling at all.
///
/// What keeps the two honest is that the server always re-prices on arrival and
/// never trusts this figure. A disagreement is recorded as a discrepancy rather
/// than silently accepted, so a drift between these two files surfaces as data
/// a person looks at instead of money quietly going missing.
///
/// Every rule below is the server's rule, for the server's reason:
///
/// **Integer cents throughout, never a double.** A double cannot hold 0.10
/// exactly. Dart's `int` is 64-bit on the VM but compiles to a JS double on the
/// web; this app is Android-only, so the arithmetic is exact.
///
/// **Discount before tax.** A discount reduces the taxable amount. Taxing first
/// would charge tax on money nobody paid.
///
/// **A cart discount is apportioned across lines before tax.** Tax is per line,
/// because each line carries its own rate and inclusive flag, so there is no
/// single rate at which a whole-cart discount could be taxed.
///
/// **Round per line, then sum.** Never round a total. Rounding the sum lets the
/// total disagree with the lines printed above it, which is exactly the thing a
/// customer checks.
library;

/// Basis points in one whole. A rate of 16% is 1600 bps.
const int kBpsDenominator = 10000;

/// The smallest coin in practical circulation, in cents.
const int kCashRoundingCents = 100;

class PricingError implements Exception {
  PricingError(this.message);
  final String message;
  @override
  String toString() => 'PricingError: $message';
}

/// Divide and round half away from zero.
///
/// Dart's `~/` truncates towards zero, and `round()` on a double reintroduces
/// the imprecision this whole file exists to avoid. Mirrors the server's
/// `round_half_up_div` including its negative branch.
int roundHalfUpDiv(int numerator, int denominator) {
  if (denominator <= 0) {
    throw PricingError('denominator must be positive');
  }
  if (numerator < 0) {
    return -((2 * -numerator + denominator) ~/ (2 * denominator));
  }
  return (2 * numerator + denominator) ~/ (2 * denominator);
}

int applyPercentage(int amountCents, int percentBps) =>
    roundHalfUpDiv(amountCents * percentBps, kBpsDenominator);

/// Tax already contained in a gross amount.
///
/// Derived by subtraction rather than computed independently, so that
/// `net + tax` is always exactly what the customer pays. Computing both and
/// hoping they add up is how a receipt ends up a cent out.
int taxFromInclusive(int grossCents, int rateBps) =>
    roundHalfUpDiv(grossCents * rateBps, kBpsDenominator + rateBps);

int taxOnExclusive(int netCents, int rateBps) =>
    roundHalfUpDiv(netCents * rateBps, kBpsDenominator);

/// Round to the smallest coin that actually circulates.
int roundCash(int amountCents, {int increment = kCashRoundingCents}) {
  if (increment <= 0) throw PricingError('increment must be positive');
  return roundHalfUpDiv(amountCents, increment) * increment;
}

/// One line as the cashier has entered it.
class LineInput {
  const LineInput({
    required this.itemId,
    required this.name,
    required this.unitPriceCents,
    required this.quantityMilli,
    this.sku = '',
    this.unit = 'EACH',
    this.taxRateBps = 0,
    this.taxIsInclusive = true,
    this.discountBps = 0,
    this.discountCents = 0,
  });

  final String itemId;
  final String name;
  final String sku;
  final String unit;
  final int unitPriceCents;

  /// Quantity in thousandths, because goods are sold by weight and 0.333 kg has
  /// to be representable exactly. An integer here for the same reason money is
  /// an integer: a double would make 2.5 kg of sugar occasionally price a cent
  /// differently than the server did.
  final int quantityMilli;

  final int taxRateBps;
  final bool taxIsInclusive;
  final int discountBps;
  final int discountCents;

  double get quantity => quantityMilli / 1000;

  /// Quantity times price, rounded once.
  ///
  /// Rounded here because quantity is fractional for goods sold by weight -
  /// 2.5 kg at KES 180 is exact, but 0.333 kg is not, and the cent has to be
  /// resolved somewhere. Doing it once, at the line, keeps every later figure
  /// exact.
  int grossBeforeDiscount() =>
      roundHalfUpDiv(unitPriceCents * quantityMilli, 1000);

  LineInput copyWith({int? quantityMilli, int? unitPriceCents, int? discountBps, int? discountCents}) =>
      LineInput(
        itemId: itemId,
        name: name,
        sku: sku,
        unit: unit,
        unitPriceCents: unitPriceCents ?? this.unitPriceCents,
        quantityMilli: quantityMilli ?? this.quantityMilli,
        taxRateBps: taxRateBps,
        taxIsInclusive: taxIsInclusive,
        discountBps: discountBps ?? this.discountBps,
        discountCents: discountCents ?? this.discountCents,
      );
}

/// One priced line.
class LineTotals {
  const LineTotals({
    required this.line,
    required this.grossBeforeDiscountCents,
    required this.lineDiscountCents,
    required this.cartDiscountShareCents,
    required this.netCents,
    required this.taxCents,
    required this.grossCents,
  });

  final LineInput line;
  final int grossBeforeDiscountCents;
  final int lineDiscountCents;
  final int cartDiscountShareCents;
  final int netCents;
  final int taxCents;
  final int grossCents;

  int get totalDiscountCents => lineDiscountCents + cartDiscountShareCents;
}

/// A whole priced cart.
class CartTotals {
  const CartTotals({
    required this.lines,
    required this.subtotalCents,
    required this.discountCents,
    required this.taxCents,
    required this.totalCents,
  });

  static const empty = CartTotals(
    lines: [],
    subtotalCents: 0,
    discountCents: 0,
    taxCents: 0,
    totalCents: 0,
  );

  final List<LineTotals> lines;
  final int subtotalCents;
  final int discountCents;
  final int taxCents;
  final int totalCents;

  int get lineCount => lines.length;

  /// What the customer is asked for in cash, rounded to the shilling.
  int get cashDueCents => roundCash(totalCents);

  /// The difference rounding made, recorded so the drawer reconciles exactly
  /// rather than drifting a few shillings a day.
  int get roundingAdjustmentCents => cashDueCents - totalCents;
}

/// Work out a discount in cents, however it was expressed.
///
/// Percentage and fixed amount may both be present, and both apply. Capped at
/// the base, because a discount larger than the thing being discounted would
/// produce a negative line and, downstream, negative tax.
int resolveDiscount(int baseCents, {int discountBps = 0, int discountCents = 0}) {
  if (discountBps < 0 || discountCents < 0) {
    throw PricingError('A discount cannot be negative.');
  }
  if (discountBps > kBpsDenominator) {
    throw PricingError('A discount cannot exceed 100%.');
  }
  final total = applyPercentage(baseCents, discountBps) + discountCents;
  final cap = baseCents > 0 ? baseCents : 0;
  return total < cap ? total : cap;
}

/// Split an amount across weights so the parts sum exactly to the whole.
///
/// Largest-remainder, ties broken by the larger weight then by index, so the
/// result is deterministic rather than dependent on iteration order. The
/// exactness is the point: a cart discount that apportions to one cent less
/// than it should leaves a total that does not match the sum of its lines, and
/// the difference gets silently absorbed into tax.
List<int> apportion(int amountCents, List<int> weights) {
  if (amountCents == 0 || weights.isEmpty) {
    return List<int>.filled(weights.length, 0);
  }
  if (weights.any((w) => w < 0)) {
    throw PricingError('Weights cannot be negative.');
  }

  final totalWeight = weights.fold<int>(0, (a, b) => a + b);
  if (totalWeight == 0) {
    // Every line is free, so there is nothing to take a discount off. Splitting
    // equally instead would create negative lines.
    return List<int>.filled(weights.length, 0);
  }

  final shares = [
    for (final weight in weights) (amountCents * weight) ~/ totalWeight,
  ];
  var remainder = amountCents - shares.fold<int>(0, (a, b) => a + b);

  if (remainder != 0) {
    final fractions = [
      for (final weight in weights) (amountCents * weight) % totalWeight,
    ];
    final order = List<int>.generate(weights.length, (i) => i)
      ..sort((a, b) {
        final byFraction = fractions[b].compareTo(fractions[a]);
        if (byFraction != 0) return byFraction;
        final byWeight = weights[b].compareTo(weights[a]);
        if (byWeight != 0) return byWeight;
        return a.compareTo(b);
      });
    for (var position = 0; position < remainder; position++) {
      shares[order[position % order.length]] += 1;
    }
  }

  return shares;
}

/// Price a whole cart, lines and all.
///
/// Line gross, then line discount, then apportion the cart discount over what
/// remains, then split tax on each line using that line's own rate and
/// inclusive flag. Mixed inclusive and exclusive lines therefore total
/// correctly on one sale, which a per-business tax setting could not express.
CartTotals priceCart(
  List<LineInput> lines, {
  int cartDiscountBps = 0,
  int cartDiscountCents = 0,
}) {
  if (lines.isEmpty) throw PricingError('A sale needs at least one line.');

  final grosses = [for (final line in lines) line.grossBeforeDiscount()];
  final lineDiscounts = [
    for (var i = 0; i < lines.length; i++)
      resolveDiscount(
        grosses[i],
        discountBps: lines[i].discountBps,
        discountCents: lines[i].discountCents,
      ),
  ];
  final afterLineDiscount = [
    for (var i = 0; i < lines.length; i++) grosses[i] - lineDiscounts[i],
  ];

  final cartDiscount = resolveDiscount(
    afterLineDiscount.fold<int>(0, (a, b) => a + b),
    discountBps: cartDiscountBps,
    discountCents: cartDiscountCents,
  );
  final shares = apportion(cartDiscount, afterLineDiscount);

  final priced = <LineTotals>[];
  var subtotal = 0, discount = 0, tax = 0, total = 0;

  for (var i = 0; i < lines.length; i++) {
    final line = lines[i];
    final charged = afterLineDiscount[i] - shares[i];

    int net, lineTax, lineGross;
    if (line.taxRateBps <= 0) {
      net = charged;
      lineTax = 0;
      lineGross = charged;
    } else if (line.taxIsInclusive) {
      lineTax = taxFromInclusive(charged, line.taxRateBps);
      net = charged - lineTax;
      lineGross = charged;
    } else {
      lineTax = taxOnExclusive(charged, line.taxRateBps);
      net = charged;
      lineGross = charged + lineTax;
    }

    priced.add(
      LineTotals(
        line: line,
        grossBeforeDiscountCents: grosses[i],
        lineDiscountCents: lineDiscounts[i],
        cartDiscountShareCents: shares[i],
        netCents: net,
        taxCents: lineTax,
        grossCents: lineGross,
      ),
    );

    subtotal += net;
    discount += lineDiscounts[i] + shares[i];
    tax += lineTax;
    total += lineGross;
  }

  return CartTotals(
    lines: priced,
    subtotalCents: subtotal,
    discountCents: discount,
    taxCents: tax,
    totalCents: total,
  );
}

/// Change owed on a cash tender.
int changeFor(int tenderedCents, int dueCents) {
  if (tenderedCents < dueCents) {
    throw PricingError('Tendered less than the amount due.');
  }
  return tenderedCents - dueCents;
}
